book/go/api/writing-your-own-stateful-application.md

```diff
@@ -167,18 +167,19 @@ That is, while the stateless computation constructor `To` took only a computatio
 
 Partitioning is a key aspect of how work is distributed in Wallaroo. From the application's point of view, what is required is:
 
-* the keys to partition the state objects
 * a partitioning function
 * partition-compatible states - the state should be defined in such a way that each partition can have its own distinct state instance
 
+The keys to partition the state objects are also important, but are optional here. A wallaroo application is able to dynamically create the keys as needed, if required. 
```

Comment:
Maybe something like: `A Wallaroo application uses the user-defined partition function to extract a key from every message that will be routed to state. If we haven't seen a key before, Wallaroo will dynamically create a new state partition corresponding to it.`